Fix layouts for mobile screens

JustinBeBoy opened the pull request:
This PR is a continuation of #963 

This PR fixs: 

- [x] When the software keyboard is showing, the hide balance icon is hidden
- [x] The "transaction" text on the bottom nav is not showing fully
- [x] When changing setings(dark mode, currency conversion, language, etc...) the bottom nav gets distorted
- [x] Fix not show all seeds word option in verify seed words page
- [x] Make 2 columns of 33 seeds word in save seed page on mobile
- [x] Fix dark mode not working when change from setting
- [x] Add margin to title and buttons in start page
- [x] Fix button "View on politeia" in proposal detail page not show
- [x] Fix "Clear all" and "validate wallet seeds" button not show when keyboard appears in restore wallet page
- [x] Close the gap to increase the visibility of information in many page
- [x] Right margin and items of ticket record are uneven in staking page

**Screenshots**

| <img src="https://user-images.githubusercontent.com/19331824/174628967-2a835c2b-eef2-452e-82bf-b8ec7e823006.png" height="500">|<img src="https://user-images.githubusercontent.com/19331824/174629055-c344f362-7602-41d7-851c-7071b7f47305.png" height="500"> |
|-|-|


| <img src="https://user-images.githubusercontent.com/19331824/174629181-64e96802-8b59-470f-bf55-b06663b11bcc.png" height="500">|<img src="https://user-images.githubusercontent.com/19331824/174629228-ddc4bb78-ee58-4a4a-9ae6-f6e12388595c.png" height="500"> |
|-|-|

| <img src="https://user-images.githubusercontent.com/19331824/174629385-dd91e4ad-b638-49d8-a2aa-3ff59d7a1835.png" height="500">|<img src="https://user-images.githubusercontent.com/19331824/174629421-e239ac2a-2f1a-4f79-8132-2d2004187802.png" height="500"> |
|-|-|

| <img src="https://user-images.githubusercontent.com/19331824/174629462-b2d4e056-61e0-4409-afb3-cd9c038d0933.png" height="500">|<img src="https://user-images.githubusercontent.com/19331824/174629577-84da3d1e-6396-4bdf-b5e6-a7f178a3f241.png" height="500"> |
|-|-|

| <img src="https://user-images.githubusercontent.com/19331824/174629592-9963e32e-c7d2-4348-af56-22da9c606fa3.png" height="500">|<img src="https://user-images.githubusercontent.com/19331824/174629598-ec1f856b-cfe4-4fcc-95c2-d24561055870.png" height="500"> |
|-|-|

| <img src="https://user-images.githubusercontent.com/19331824/174629600-3d884f2e-90d0-403e-9959-4f901a263cd6.png" height="500">|<img src="https://user-images.githubusercontent.com/19331824/174629942-4f822f20-d03e-4109-aaaa-7fa7c1e6058c.png" height="500"> |
|-|-|

dreacot committed:
fix excess space after the nav bar

dreacot committed:
add FAB for send a receive

dreacot committed:
hide FAB on pages not overview/transactions

dreacot committed:
add click functionality to FAB

dreacot committed:
add staking to bottom nav and more menu

dreacot committed:
properly define screen width

dreacot committed:
fix mobile view for more pae

dreacot committed:
fix excesses page at the top of security tools, help and about pages
layout receive page for mobile

dreacot committed:
add global function to track realtime app width as it changes

dreacot committed:
remove hoverabe option from FAB

dreacot committed:
fix transaction page layout

dreacot committed:
make restore page suitable for mobile

JustinBeBoy committed:
update value when change setting, change inset bottom bar, fix nav bar get distorted when show keyboard

JustinBeBoy committed:
fix not show all option in seed word page by add scroll
change 3 columns to 2 columns for mobile in step 1 seed page

JustinBeBoy committed:
fix conflict

JustinBeBoy committed:
fix dark mode not work

JustinBeBoy committed:
refactor code

dreacot commented on ui/page/seedbackup/save_seed.go:
```diff
@@ -90,6 +91,22 @@ func (pg *SaveSeedPage) OnNavigatedTo() {
 
+				//for mobile
+				rowm1 := wordList[:17]
+				rowm2 := wordList[17:]
```
```suggestion
				rowMobile1 := wordList[:17]
				rowMobile2 := wordList[17:]
```

JustinBeBoy replied:
I have change DefualtThemeColors() function with its true purpose, I will return new object of color, it shouldn't change current color because some function in dark mode use it to get the default color

dreacot commented on ui/values/colors.go:
```diff
@@ -71,54 +71,56 @@ func (c *Color) DarkThemeColors() {
-	c.Primary = rgb(0x2970ff)
-	c.Primary50 = rgb(0xE3F2FF)
-	c.PrimaryHighlight = rgb(0x1B41B3)
+func (c *Color) DefaultThemeColors() *Color {
+	cl := Color{
+		Primary:          rgb(0x2970ff),
+		Primary50:        rgb(0xE3F2FF),
+		PrimaryHighlight: rgb(0x1B41B3),
```
why this change, what aren't we using `c`

JustinBeBoy replied:
DefaultThemeColors function should to get default color, it shouldn't a set color function because some page use this function for get default color

dreacot reviewed (changes requested):
Left some comments, other than that, this looks good

JustinBeBoy committed:
fix conflict

JustinBeBoy committed:
layout start page for mobile

JustinBeBoy committed:
fix buttons in restore page not show when keyboard appears

JustinBeBoy committed:
fix layout to optimize setting page, log page, statistic page, license page for mobile

JustinBeBoy committed:
- Fix Uniform for mobile and update page used
- Fix layout to optimize display more page, security tool page, validate address page, verify message page, proposal page, overview page, transaction page, wallet page for mobile

JustinBeBoy committed:
- fix layout to optimize display consensus page, save seed page, verify seed page, account detail page, sign message page, wallet setting page for mobile

JustinBeBoy committed:
fix layout to optimize display about page, debug page, help page, receive page, send page, proposal detal page for mobile
fix button "View on politeia" in proposal detail page not show

JustinBeBoy committed:
validate to fix crash when change to staking tab

JustinBeBoy committed:
- Fix layout to optimize display staking page for mobile

JustinBeBoy committed:
set max size for items in ticket record section

JustinBeBoy committed:
fix layout to optimize display transaction detail page, account mixer page, staking list page for mobile

JustinBeBoy committed:
Merge branch 'master' into botton_bar_fix

# Conflicts:
#	ui/page/governance/governance_page.go
#	ui/page/seedbackup/backup_instructions.go
#	ui/page/seedbackup/save_seed.go
#	ui/page/seedbackup/verify_seed.go
#	ui/page/start_page.go
#	ui/page/wallets/restore_page.go

dreacot commented on ui/page/seedbackup/verify_seed.go:
```diff
@@ -40,7 +40,7 @@ type VerifySeedPage struct {
 	backButton   decredmaterial.IconButton
 	actionButton decredmaterial.Button
-	seedList     *layout.List
+	listSeedList []*layout.List
```
this is not directly clear to me what this does, we could use a ore appropriate name, or add a comment

dreacot reviewed (changes requested):
Overall this looks great. some minor touch ups and i'd merge

1. The button at the bottom right doesn't have the correct bottom and right margins
<img src="https://user-images.githubusercontent.com/25265396/175100770-1f8aee72-c1e9-4fc8-8799-a5d0ca7a5bfb.png" width="500"/>

2. The floating send and receive button, doesn't have the correct the correct bottom margin
<img src="https://user-images.githubusercontent.com/25265396/175100827-27a65e47-b813-4871-8240-e66b242b3242.png" width="500"/>

3. This page doesn't have the proper left and right margins
<img src="https://user-images.githubusercontent.com/25265396/175101156-62392706-051e-4748-a2dc-3386fa9c5cd9.png" width="500"/>

4. this item can be made full width
<img src="https://user-images.githubusercontent.com/25265396/175101230-4de543b1-4787-415a-b728-d394d870c0b2.png" width="500"/>

Sirmorrison commented on ui/decredmaterial/theme.go:
```diff
@@ -74,9 +74,8 @@ func NewTheme(fontCollection []text.FontFace, decredIcons map[string]image.Image
 
 func (t *Theme) SwitchDarkMode(isDarkModeOn bool, decredIcons map[string]image.Image) {
-	t.Color.DefualtThemeColors()
+	t.Color = t.Color.DefaultThemeColors()
```
I dont see the reason for this change. if it does not affect your code in anyway, kindly revert.

JustinBeBoy replied:
The DefualtThemeColors function should return object of default colors, it shouldn't change current colors of theme because some page in app use this function to get default color

dreacot replied:
@JustinBeBoy perhaps you can create a new issue for this, and have it solved in a separate PR, properly documenting the reason for the change. As the added change does not affect this PR/ is not relating to this PR, it's outside the scope of this PR.

JustinBeBoy replied:
It's not outside the scope of this PR because it is the cause of problems with dark mode

dreacot replied:
oh I see, thanks for pointing that out, i thought it was just some code optimization

Sirmorrison commented on ui/decredmaterial/progressbar.go:
```diff
@@ -152,7 +152,10 @@ func (mp *MultiLayerProgressBar) progressBarLayout(gtx C) D {
 
 	calProgressWidth := func(progress int) int {
-		val := (progress / mp.total) * 100
+		val := 1
+		if mp.total != 0 {
+			val = (progress / mp.total) * 100
+		}
```
Why are we making this update?

JustinBeBoy replied:
in my case for some reason mp.total = 0 causes the app to crash when switching to staking tab, it should be checked before doing the division

dreacot replied:
i think converting to float should fix the crash

Sirmorrison commented on ui/page/components/components.go:
```diff
@@ -108,6 +108,31 @@ func UniformHorizontalPadding(gtx layout.Context, body layout.Widget) layout.Dim
+		Right: padding,
+		Left:  values.MarginPadding10,
+	}.Layout(gtx, body)
+}
```
The only difference i see with the 2 methods is the top inset. this can be merged into 1 and the inset passed as a parameter.

```suggestion
func UniformMobile(gtx layout.Context, isHorizontal, withList bool, body layout.Widget) layout.Dimensions {

	insetRight := values.MarginPadding10
	if withList {
		insetRight = values.MarginPadding0
	}
	
	insetTop := values.MarginPadding24
	if isHorizontal{
	     insetTop = values.MarginPadding0
	}
	
	return layout.Inset{
		Top:   insetTop,
		Right: insetRight,
		Left:  values.MarginPadding10,
	}.Layout(gtx, body)
}
```

Sirmorrison commented on ui/page/debug_page.go:
```diff
@@ -162,7 +162,18 @@ func (pg *DebugPage) Layout(gtx C) D {
+}
+
+func (pg *DebugPage) layoutMobile(gtx layout.Context, body layout.Widget) layout.Dimensions {
+	return components.UniformMobile(gtx, false, body)
```
I dont see the need for the `layoutDesktop` and `layoutMobile` functions as they are only used once.

```suggestion
	if pg.Load.GetCurrentAppWidth() <= gtx.Dp(values.StartMobileView) {
               // mobile layout
		return components.UniformMobile(gtx, false, container)
	}
	// desktop layout
	return components.UniformPadding(gtx, container) 
}
	
```

dreacot replied:
we should leave as is, further details here https://github.com/planetdecred/godcr/pull/971#discussion_r904149266

Sirmorrison replied:
Then i believe it wont be out of order to properly comments these codes for easier reviews.

dreacot replied:
yeah, your right

Sirmorrison commented on ui/page/governance/consensus_page.go:
```diff
@@ -275,6 +282,73 @@ func (pg *ConsensusPage) Layout(gtx C) D {
+					// 	gtx.Constraints.Max.X = gtx.Dp(values.MarginPadding150)
+					// 	gtx.Constraints.Min.X = gtx.Constraints.Max.X
+
+					//TODO: temp removal till after V1
+					// card := pg.Theme.Card()
+					// card.Radius = decredmaterial.Radius(8)
+					// return card.Layout(gtx, func(gtx C) D {
+					// 	return layout.Inset{
+					// 		Left:   values.MarginPadding10,
+					// 		Right:  values.MarginPadding10,
+					// 		Top:    values.MarginPadding2,
+					// 		Bottom: values.MarginPadding2,
+					// 	}.Layout(gtx, pg.searchEditor.Layout)
+					// })
+					// }),
+					layout.Expanded(func(gtx C) D {
+						gtx.Constraints.Min.X = gtx.Constraints.Max.X
+						return layout.E.Layout(gtx, func(gtx C) D {
+							card := pg.Theme.Card()
+							card.Radius = decredmaterial.Radius(8)
+							return layout.Inset{Right: values.MarginPadding10}.Layout(gtx, func(gtx C) D {
+								return card.Layout(gtx, func(gtx C) D {
+									return layout.UniformInset(values.MarginPadding8).Layout(gtx, func(gtx C) D {
+										return pg.layoutSyncSection(gtx)
+									})
+								})
+							})
+						})
+					}),
+					layout.Expanded(func(gtx C) D {
+						return pg.orderDropDown.Layout(gtx, 55, true)
+					}),
+					layout.Expanded(func(gtx C) D {
+						return pg.walletDropDown.Layout(gtx, pg.orderDropDown.Width+51, true)
+					}),
+				)
+			})
+		}),
+	)
+}
```
I dont see the difference between what you have here and on the desktop layout.. Kindly help clarify why we need this?

Actually lookign again i see the dropdown position values changed if that be the only case, we can have only on function that takes the dropdown position as a parameter and the values can be passed to it by the different layouts.

```
 func (pg *ConsensusPage) pageLayout(gtx layout.Context, orderPos int, walletPos int) layout.Dimensions {}
```

```
if pg.Load.GetCurrentAppWidth() <= gtx.Dp(values.StartMobileView) {
		return pg.pageLayout(gtx, val, val)
	}
   return pg.pageLayout(gtx, val, val)
```

dreacot replied:
we should leave as is, further details here https://github.com/planetdecred/godcr/pull/971#discussion_r904149266

Sirmorrison replied:
Then i believe it wont be out of order to properly comments these codes for easier reviews

Sirmorrison commented on ui/page/governance/governance_page.go:
```diff
@@ -117,6 +124,24 @@ func (pg *Page) Layout(gtx C) D {
+			}),
+		)
+	})
+}
```
No need for this as i see the only use was is in the splash screen `uniformMobileLayout()`

dreacot replied:
we should leave as is, further details here https://github.com/planetdecred/godcr/pull/971#discussion_r904149266

Sirmorrison commented on ui/page/governance/governance_page.go:
```diff
@@ -99,6 +99,13 @@ func (pg *Page) HandleUserInteractions() {
 }
 
 func (pg *Page) Layout(gtx C) D {
+	if pg.Load.GetCurrentAppWidth() <= gtx.Dp(values.StartMobileView) {
+		return pg.layoutMobile(gtx)
+	}
+	return pg.layoutDesktop(gtx)
+}
+
+func (pg *Page) layoutDesktop(gtx layout.Context) layout.Dimensions {
 	if !pg.isGovernanceFeatureEnabled() {
 		return components.UniformPadding(gtx, pg.splashScreenLayout)
 	}
```
```suggestion

if !pg.isGovernanceFeatureEnabled() {
	if pg.Load.GetCurrentAppWidth() <= gtx.Dp(values.StartMobileView) {
	    return components.UniformMobile(gtx, false, pg.splashScreenLayout)
	}
		return components.UniformPadding(gtx, pg.splashScreenLayout)
}
```

dreacot replied:
we should leave as is, further details here https://github.com/planetdecred/godcr/pull/971#discussion_r904149266

Sirmorrison commented on ui/page/governance/proposal_details_page.go:
```diff
@@ -579,3 +586,59 @@ func (pg *ProposalDetails) Layout(gtx C) D {
 	return components.UniformPadding(gtx, body)
 }
+
+func (pg *ProposalDetails) layoutMobile(gtx layout.Context) layout.Dimensions {
+	proposal := pg.proposal
+	_, ok := pg.proposalItems[proposal.Token]
+	if !ok && !pg.loadingDescription {
+		pg.loadingDescription = true
+		go func() {
+			var proposalDescription string
+			if proposal.IndexFile != "" && proposal.IndexFileVersion == proposal.Version {
+				proposalDescription = proposal.IndexFile
+			} else {
+				var err error
+				proposalDescription, err = pg.WL.MultiWallet.Politeia.FetchProposalDescription(proposal.Token)
+				if err != nil {
+					log.Errorf("Error loading proposal description: %v", err)
+					time.Sleep(7 * time.Second)
+					pg.loadingDescription = false
+					return
```
Same review above applies here.

dreacot replied:
we should leave as is, further details here https://github.com/planetdecred/godcr/pull/971#discussion_r904149266

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -264,6 +271,57 @@ func (pg *ProposalsPage) Layout(gtx C) D {
 	)
 }
 
+func (pg *ProposalsPage) layoutMobile(gtx layout.Context) layout.Dimensions {
+	return layout.Flex{Axis: layout.Vertical}.Layout(gtx,
+		layout.Rigid(func(gtx layout.Context) layout.Dimensions {
+			return layout.Inset{Right: values.MarginPadding10}.Layout(gtx, pg.layoutSectionHeader)
+		}),
+		layout.Flexed(1, func(gtx C) D {
+			return layout.Inset{Top: values.MarginPadding10}.Layout(gtx, func(gtx C) D {
+				return layout.Stack{}.Layout(gtx,
+					layout.Expanded(func(gtx C) D {
+						return layout.Inset{Top: values.MarginPadding60}.Layout(gtx, pg.layoutContent)
+					}),
+					//TODO: temp removal till after V1
+					// layout.Expanded(func(gtx C) D {
+					// 	gtx.Constraints.Max.X = gtx.Dp(values.MarginPadding150)
+					// 	gtx.Constraints.Min.X = gtx.Constraints.Max.X
+
+					// 	card := pg.Theme.Card()
+					// 	card.Radius = decredmaterial.Radius(8)
+					// 	return card.Layout(gtx, func(gtx C) D {
+					// 		return layout.Inset{
+					// 			Left:   values.MarginPadding10,
+					// 			Right:  values.MarginPadding10,
+					// 			Top:    values.MarginPadding2,
+					// 			Bottom: values.MarginPadding2,
+					// 		}.Layout(gtx, pg.searchEditor.Layout)
+					// 	})
+					// }),
+					layout.Expanded(func(gtx C) D {
+						gtx.Constraints.Min.X = gtx.Constraints.Max.X
+						return layout.E.Layout(gtx, func(gtx C) D {
+							card := pg.Theme.Card()
+							card.Radius = decredmaterial.Radius(8)
+							return layout.Inset{Right: values.MarginPadding10}.Layout(gtx, func(gtx C) D {
+								return card.Layout(gtx, func(gtx C) D {
+									return layout.UniformInset(values.MarginPadding8).Layout(gtx, pg.layoutSyncSection)
```
These mobile layouts are just a bunch or repeated codes. Lets have them optimized as mentioned above

dreacot replied:
For now they might look like they are, but they are not. currently we are trying to share the same layouts for desktop and mobile, and making some optimizations to the code in order to make it look somewhat presentable on mobile. However, we are going to be implementing totally different UIs for mobile further down the line when proper mockups become available. The mobile and desktop views need to be clearly distinguishable.

This PR, the ones before it, and the ones coming after it, are aimed at setting the bedrock foundation for the godcr mobile builds

Having said that the current layout separation is in line with the scope of this PR. We should leave as is.

Sirmorrison replied:
Then i believe it wont be out of order to properly comments these codes for easier reviews.

Sirmorrison reviewed (dismissed):
These mobile layouts are just a bunch or repeated codes with little to not significant updates. Lets have them optimized as mentioned in the comments.

JustinBeBoy committed:
- Fix layout to optimize display verify seed page for mobile, update layout in wallet page, main page, help page
- Optimize UniformMobile function

dreacot commented on ui/page/components/components.go:
```diff
@@ -108,27 +108,31 @@ func UniformHorizontalPadding(gtx layout.Context, body layout.Widget) layout.Dim
+// 		Right: padding,
+// 		Left:  values.MarginPadding10,
+// 	}.Layout(gtx, body)
+// }
```
let's remove this comment if it's not needed

dreacot reviewed (changes requested):
align the text and icon left, put them in same row
<img src="https://user-images.githubusercontent.com/25265396/175520743-286b843c-ee19-43c4-ab72-2b0aa6cdf113.png" height="500" />

just like we have here
<img src="https://user-images.githubusercontent.com/25265396/175520765-5c825c5f-5733-403a-98ab-1a3ec738caf6.png" height="500" />

JustinBeBoy committed:
remove function not need

JustinBeBoy committed:
update item in help page

JustinBeBoy committed:
Merge branch 'botton_bar_fix' of jtd.github.com:JustinBeBoy/godcr into botton_bar_fix

dreacot reviewed (approved):
LGTM
